### DISCUSSION TITLE
Required input struct generated as pointer

### DISCUSSION THREAD

**Discussion by Junaid-Ahm** [99designs/gqlgen#2156] [Q&A]
### What happened?

For a required object field in another schema object (happens with both `type` or `input`), it declares the required object field as pointer to the struct - see field `Details` declared as `*UserDetailsInput` in declaration of `UpdateUserInput`. It should have been declared as `UserDetailsInput` instead.

The `models_gen.go` contains

```
type UpdateUserInput struct {
	ID      string            `json:"id"`
	Name    string            `json:"name"`
	Details *UserDetailsInput `json:"details"`
}

type UserDetailsInput struct {
	X string  `json:"x"`
	Y string  `json:"y"`
	Z *string `json:"z"`
}
```

### What did you expect?
If the object field is required, it should be declared without pointer. In other words, it should only be declared as pointer if the object field is optional.

The `models_gen.go` should contain

```
type UpdateUserInput struct {
	ID      string           `json:"id"`
	Name    string           `json:"name"`
	Details UserDetailsInput `json:"details"`
}
```

### Minimal graphql.schema and models to reproduce

```
type Mutation {
    updateUser(input: UpdateUserInput!): User!
}

input UserDetailsInput {
    x: String!
    y: String!
    z: String
}

input UpdateUserInput {
    id: ID!
    name: String!
    details: UserDetailsInput!
}
```

### versions
 - `go run github.com/99designs/gqlgen version`: v0.17.4
 - `go version`: 1.18

**Comment by frederikhors** [comment]
Already discussed hundreds of times. Search through issues and discussions.

**Reply by ianling** [reply]
@frederikhors I searched and could not find anything related to this specific issue. Can you link directly to a past discussion of this? I am running into the same behavior and I don't understand why it is designed this way.

**Reply by ianling** [reply]
Found it! https://github.com/99designs/gqlgen/pull/701

To summarize, it seems this was due to the need to check for cyclical relationships between structs. It was quicker and easier to simply turn all structs within structs into pointers. The more complicated approach would have been to try to detect cyclical relationships by traversing the struct's fields, but it's understandable that the "always pointers" approach was taken instead, as it avoids any potential bugs in the struct traversal code.

**Reply by ianling** [reply]
I add a config option for this in https://github.com/99designs/gqlgen/pull/2174
